ethereum/contracts/query/QueryResponse.sol

```diff
@@ -0,0 +1,246 @@
+// contracts/query/QueryResponse.sol
+// SPDX-License-Identifier: Apache 2
+
+pragma solidity ^0.8.0;
+
+import {BytesParsing} from "../relayer/libraries/BytesParsing.sol";
+import "../interfaces/IWormhole.sol";
+
+// @dev ParsedQueryResponse is returned by QueryResponse.parseAndVerifyQueryResponse().
+struct ParsedQueryResponse {
+    uint8   version;
+    uint16  senderChainId;
+    uint32  nonce;
+    bytes   requestId; // 65 byte sig for off-chain, 32 byte vaaHash for on-chain
+    ParsedPerChainQueryResponse [] responses;
+}
+
+// @dev ParsedPerChainQueryResponse describes a single per-chain response.
+struct ParsedPerChainQueryResponse {
+    uint16 chainId;
+    uint8 queryType;
+    bytes request;
+    bytes response;
+}
+
+// @dev EthCallQueryResponse describes an ETH call per-chain query.
+struct EthCallQueryResponse {
+    bytes requestBlockId;
+    uint64 blockNum;
+    uint64 blockTime;
+    bytes32 blockHash;
+    EthCallData [] result;
+}
+
+// @dev EthCallData describes a single ETH call query / response pair.
+struct EthCallData {
+    address contractAddress;
+    bytes callData;
+    bytes result;
+}
+
+// Custom errors
+error InvalidResponseVersion();
+error VersionMismatch();
+error NumberOfResponsesMismatch();
+error ChainIdMismatch();
+error RequestTypeMismatch();
+error UnsupportedQueryType();
+error UnexpectedNumberOfResults();
+error InvalidPayloadLength(uint256 received, uint256 expected);
+
+// @dev QueryResponse is a library that implements the parsing and verification of Cross Chain Query (CCQ) responses.
+abstract contract QueryResponse {
+    using BytesParsing for bytes;
+
+    bytes public constant responsePrefix = bytes("query_response_0000000000000000000|");
+    uint8 public constant QT_ETH_CALL = 1;
+
+    /// @dev getResponseHash computes the hash of the specified query response.
+    function getResponseHash(bytes memory response) public pure returns (bytes32) {
+        return keccak256(response);
+    }
+
+    /// @dev getResponseDigest computes the digest of the specified query response.
+    function getResponseDigest(bytes memory response) public pure returns (bytes32) {
+        return keccak256(abi.encodePacked(responsePrefix,getResponseHash(response)));
+    }
+    
+    /// @dev parseAndVerifyQueryResponse verifies the query response and returns the parsed response.
+    function parseAndVerifyQueryResponse(address wormhole, bytes memory response, IWormhole.Signature[] memory signatures) public view returns (ParsedQueryResponse memory r) {
+        verifyQueryResponseSignatures(wormhole, response, signatures);
+
+        uint index = 0;
+        
+        (r.version, index) = response.asUint8Unchecked(index);
+        if (r.version != 1) {
+            revert InvalidResponseVersion();
+        }
+
+        (r.senderChainId, index) = response.asUint16Unchecked(index);
+
+        if (r.senderChainId == 0) {
```

Comment:
A comment would be helpful here (instead of having to look at the `parsedQueryResponse` struct.